File: server/apps/users/services.py
from os import name
from config.supabase import supabase

def register(email:str, password: str, username:str):
    
    existing_username = supabase.table("Profile").select("name").eq("name", username).execute()

    if len(existing_username.data) > 0:
        raise ValueError("Username is already taken")

    response =  supabase.auth.sign_up({
        "email": email,
        "password": password,
        "options": {
            "data": {
                "username": username
            }
        }
    })
        
    return response.user.email

def login(email:str, password:str):
    response = supabase.auth.sign_in_with_password({
        "email": email,
        "password": password
    })
    return response.session

def verifyOtp(email:str, otp:str):
    response = supabase.auth.verify_otp({
        "email": email,
        "token": otp,
        "type": "email"
    })
    if response.session:
        # Extract the username we saved earlier
        user_metadata = response.user.user_metadata
        username = user_metadata.get("username", "Unknown")
        
        # Now that they have a session, this insert will work!
        supabase.table("Profile").insert({
            "id": response.user.id,
            "name": username
        }).execute()

    return response.user.email

File: server/src/apps/users/services.js
import { supabaseAdmin } from "../../config/supabase.js";

export const refresh = async (token) => {
    const { data, error } = await supabaseAdmin.auth.refreshSession({
        refresh_token: token
    })

    if (error) {
        console.error("Services Error: Refresh, ", error)
        throw new Error(error)
    }
    return data
}

export const register = async (username, email, password) => {

    const { data: existingUser } = supabaseAdmin
        .from("Profiles")
        .select("name")
        .eq("name", username)
        .single()

    if (existingUser) {
        console.error("Services error: Username already taken")
        throw new Error("Username is already taken.")
    }
    const { data, error } = await supabaseAdmin.auth.signUp({
        email,
        password,
        options: {
            data: {
                username: username
            },
        }
    })

    if (error) throw error;
    return data
}

export const verifyRegister = async (email, token) => {

    const { data, error: verifyError } = await supabaseAdmin.auth.verifyOtp({
        email: email,
        token: token,
        type: 'email',

    })
    console.log(data)
    const { error: ProfileError } = await supabaseAdmin
        .from("Profile")
        .insert({
            id: data?.user?.id,
            name: data.user.user_metadata.username,
        })

    if (ProfileError) {
        console.error("Services Error: Profile, ", ProfileError)
        throw new Error(ProfileError)
    };
    if (verifyError) {
        console.error("Services Error: Verify, ", verifyError)
        throw new Error(verifyError)
    };

    return data
}

export const login = async (email, password) => {
    const { data, error } = await supabaseAdmin.auth.signInWithPassword({
        email, password
    })

    if (error) {
        console.error("Services Error: Login, ", error.message)
        throw new Error(error.message)
    }

    return data
}